File: framework/pages/automationexercise/AutomationExerciseProductDetailPage.ts
import { expect } from '@playwright/test';
import { AutomationExerciseBasePage } from './AutomationExerciseBasePage';

/**
 * @pageIdentity AutomationExerciseProductDetailPage
 * @urlPattern https://automationexercise.com/product_details/
 */
export class AutomationExerciseProductDetailPage extends AutomationExerciseBasePage {
  public async assertProductNameVisible(name: string | RegExp): Promise<void> {
    await expect(this.page.locator('.product-information h2')).toHaveText(name);
  }

  public async assertCategoryVisible(category: string | RegExp): Promise<void> {
    await expect(this.page.locator('.product-information p').filter({ hasText: category })).toBeVisible();
  }

  public async assertPriceVisible(price: string | RegExp): Promise<void> {
    // Strict mode: scope to the price span that contains only the price (not the one with Quantity)
    // The price is usually in a <span> that contains only the price text, not the one with Quantity
    await expect(
      this.page.locator('.product-information span').filter({ hasText: price }).filter({ has: this.page.locator(`:scope:has-text("${typeof price === 'string' ? price : ''}")`) }).filter({ hasNot: this.page.locator('span:has-text("Quantity")') })
    ).toBeVisible();
  }

  public async assertAvailabilityVisible(status: string | RegExp): Promise<void> {
    await expect(this.page.locator('.product-information p').filter({ hasText: /Availability:/i })).toContainText(status);
  }

  public async assertConditionVisible(condition: string | RegExp): Promise<void> {
    await expect(this.page.locator('.product-information p').filter({ hasText: /Condition:/i })).toContainText(condition);
  }

  public async assertBrandVisible(brand: string | RegExp): Promise<void> {
    // Strict mode: scope to the product-information region and filter for the Brand line
    await expect(
      this.page.locator('.product-information p').filter({ hasText: /Brand:/i }).filter({ hasText: brand })
    ).toBeVisible();
  }
}

File: framework/pages/automationexercise/automation_exercise_product_detail_page.py
import re

from playwright.sync_api import Page, expect

from framework.pages.automationexercise.automation_exercise_base_page import (
    AutomationExerciseBasePage,
)


class AutomationExerciseProductDetailPage(AutomationExerciseBasePage):
    """@pageIdentity AutomationExerciseProductDetailPage
    @urlPattern https://automationexercise.com/product_details/
    """

    def __init__(self, page: Page) -> None:
        super().__init__(page)

    def assert_product_name_visible(self, name) -> None:
        expect(self.page.locator(".product-information h2")).to_have_text(name)

    def assert_category_visible(self, category) -> None:
        expect(
            self.page.locator(".product-information p").filter(has_text=category)
        ).to_be_visible()

    def assert_price_visible(self, price) -> None:
        expect(
            self.page.locator(".product-information span").filter(has_text=price).first
        ).to_be_visible()

    def assert_availability_visible(self, status) -> None:
        expect(
            self.page.locator(".product-information p").filter(
                has_text=re.compile("Availability:", re.I)
            )
        ).to_contain_text(status)

    def assert_condition_visible(self, condition) -> None:
        expect(
            self.page.locator(".product-information p").filter(
                has_text=re.compile("Condition:", re.I)
            )
        ).to_contain_text(condition)

    def assert_brand_visible(self, brand) -> None:
        expect(
            self.page.locator(".product-information p")
            .filter(has_text=re.compile("Brand:", re.I))
            .filter(has_text=brand)
        ).to_be_visible()
